active/0051-strategies.md

```diff
@@ -0,0 +1,248 @@
+fileSystemResolver.resolve('ember-data') => `some-path/addon-modules/modules/ember-data/index.js`
+fileSystemResolver.resolve('ember-data/-private') => `some-path/addon-modules/modules/-private.js`
+```
+
+This effort could be broken down into several phases:
+
++ dead modules elimination inside of the `addons/` (application would be the main entry point and unused modules are removed only from `addons/`)
++ dead modules elimination inside of the `app/`
+  + removing unused components and helpers (requires analysing templates)
+  + removing unused initializers/services (this likely entails work on dependency injection layer as we would need access to a resolver resolution map)
++ treeshaking (Rollup-like treeshaking where we include *only* the code that is used)
+
+`Linker` would be able to take an `exclude` list of modules as a parameter. Although, valuable in some situations, it should be clearly marked as advanced API. It should be used as a last resort and serve as an "escape hatch". 
+
+It would make sense to implement `Linker` as a strategy. Developers would be able to "opt in"/"opt out" of optimising behaviour.
+
+## Deprecating `app.import` API
+
+Ember applications which choose to use `Linker` strategy should be able to remove usages of `app.import`.
+
+## Tools
+
+With growing complexity of Ember applications, it is crucial to provide more insights into final assets.
+
+Main goals are:
+
+- report raw/uglified/compressed asset sizes; [broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser)
+- find source code duplication across your javascript assets (enables you to
+  fine tune code splitting parameters to reduce bundle invalidation rates as well
+  as improve repeat page load performance)
+
+# How We Teach This
+
+This is a backward compatible change to the existing Ember CLI eco system. In order to teach users how to use strategies, we need to update the API docs with a section for this and the best practices of when to use this. A more general purpose blog post could be beneficial as well.
+
+# Drawbacks
+
+- More libraries to maintain
+- Increasing public API surface
+
+# Alternatives
+
+[Webpack](https://webpack.js.org) has become very popular in solving this similar problem. However it is unclear how WebPack could be used with Ember as Ember has both explicit and implicit dependencies. This is why we feel we need something that is aware of how Ember apps are assembled and take advantage of existing tools when it is makes sense. Since we have an opinionated way of structuring our apps, we can avoid the "wall of configuration" that other asset packaging systems are susceptible to.
+
+# Unresolved questions
+
++ Will there be an ordering problem (strategies)?
++ Will it increase build time?
++ Which strategies are going to be supported out-of-the-box?
++ Should we introduce the same API on add-on level? 
++ Will simple extension strategy “just work”?
++ Communication between strategies
+
```

Comment:
How would we verify that a given strategy (or set of strategies) produces a valid application output?

Reply:
Good question. We don't validate application output right now (usually build fails and that's how you know that something went wrong). We definitely need to provide an easy way to debug strategies which would help address some of the build issues. My assumption is that a "valid" output would depend on the application itself and suspect we need more experimentation on that front.